From d8a6f420a07181f5949d8e401a35b1d1d393e302 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Thu, 26 Nov 2015 17:35:13 +0800
Subject: [PATCH 0687/5242] MLK-11896-3 usb: chipidea: imx: add NULL check for
 usbmisc_data

commit  2fa1f7fc38317100396159ebd4e742ffcf45ee00 from
https://source.codeaurora.org/external/imx/linux-imx.git

Check if data->usbmisc_data is NULL before dereference it in probe.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <jun.li@freescale.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/ci_hdrc_imx.c |   14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/usb/chipidea/ci_hdrc_imx.c b/drivers/usb/chipidea/ci_hdrc_imx.c
index 4ad4a5f..5f12ae1 100644
--- a/drivers/usb/chipidea/ci_hdrc_imx.c
+++ b/drivers/usb/chipidea/ci_hdrc_imx.c
@@ -534,7 +534,8 @@ static int ci_hdrc_imx_probe(struct platform_device *pdev)
 	if (pdata.flags & CI_HDRC_SUPPORTS_RUNTIME_PM)
 		data->supports_runtime_pm = true;
 
-	if (data->usbmisc_data->index > 1 && (imx_has_hsic_con(data))) {
+	if (data->usbmisc_data && data->usbmisc_data->index > 1
+	    && (imx_has_hsic_con(data))) {
 		pdata.flags |= CI_HDRC_IMX_IS_HSIC;
 		data->hsic_pad_regulator = devm_regulator_get(&pdev->dev,
 									"pad");
@@ -562,7 +563,7 @@ static int ci_hdrc_imx_probe(struct platform_device *pdev)
 		}
 	}
 
-	if (of_find_property(np, "fsl,anatop", NULL)) {
+	if (of_find_property(np, "fsl,anatop", NULL) && data->usbmisc_data) {
 		data->anatop = syscon_regmap_lookup_by_phandle(np,
 							"fsl,anatop");
 		if (IS_ERR(data->anatop)) {
@@ -571,11 +572,11 @@ static int ci_hdrc_imx_probe(struct platform_device *pdev)
 			ret = PTR_ERR(data->anatop);
 			goto disable_hsic_regulator;
 		}
-		if (data->usbmisc_data)
-			data->usbmisc_data->anatop = data->anatop;
+		data->usbmisc_data->anatop = data->anatop;
 	}
 
-	if (of_find_property(np, "imx-usb-charger-detection", NULL)) {
+	if (of_find_property(np, "imx-usb-charger-detection", NULL) &&
+							data->usbmisc_data) {
 		data->imx_usb_charger_detection = true;
 		data->charger.dev = &pdev->dev;
 		data->usbmisc_data->charger = &data->charger;
@@ -618,7 +619,8 @@ static int ci_hdrc_imx_probe(struct platform_device *pdev)
 	}
 
 	/* usbmisc needs to know dr mode to choose wakeup setting */
-	data->usbmisc_data->available_role =
+	if (data->usbmisc_data)
+		data->usbmisc_data->available_role =
 			ci_hdrc_query_available_role(data->ci_pdev);
 
 	if (data->supports_runtime_pm) {
-- 
1.7.9.5

